### PR TITLE
8328741: serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java failed with unexpected owner

### PR DESCRIPTION
**Pull request by sspitsyn**
This is the test issue. The `WaitingPT3` thread posted the `MonitorWait` event but has not released the `lockCheck` monitor yet. It has been fixed to wait for each `WaitingTask` thread to really reach the `WAITING` state. The same approach is used for `EnteringTask` threads. It has been fixed to wait for each `EnteringTask` thread to reach the `BLOCKED_ON_MONITOR` state.

Testing:
 - Reran the test `serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java` locally
 - TBD: submit the mach5 tiers 1-6 as well

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328741](https://bugs.openjdk.org/browse/JDK-8328741): serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java failed with unexpected owner (**Bug** - P4)

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [68cdd8c7](https://git.openjdk.org/jdk/pull/18778/files/68cdd8c7b3d6cc9683801683f86df524e4c0911d)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18778/head:pull/18778` \
`$ git checkout pull/18778`

Update a local copy of the PR: \
`$ git checkout pull/18778` \
`$ git pull https://git.openjdk.org/jdk.git pull/18778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18778`

View PR using the GUI difftool: \
`$ git pr show -t 18778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18778.diff">https://git.openjdk.org/jdk/pull/18778.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18778#issuecomment-2056614988)

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp [hunk 1]
```diff
@@ -225,32 +177,26 @@ Java_ObjectMonitorUsage_setTestedMonitor(JNIEnv *jni, jclass cls, jobject monito
...
-  check_jvmti_status(jni, err, "setTestedMonitor: error in JVMTI SetEventNotificationMode #3");
-
-  err = jvmti->SetEventNotificationMode(event_mode, JVMTI_EVENT_MONITOR_WAITED, nullptr);
-  check_jvmti_status(jni, err, "setTestedMonitor: error in JVMTI SetEventNotificationMode #4");
 }
 
-JNIEXPORT jint JNICALL
-Java_ObjectMonitorUsage_waitsToEnter(JNIEnv *jni, jclass cls) {
+static void wait_for_state(JNIEnv *jni, jthread thread, jint exp_state) {
   RawMonitorLocker rml(jvmti, jni, event_lock);
```
Comment (lmesnik) [inline test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp:183]: The event_lock name is misleading, there are no events anymore.
Also, I am not sure if this lock is needed at all. How it is used? 

I think that wait_for_state is a good candidate to be added in the library. With some additional doc about states.

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/libObjectMonitorUsage.cpp]: Good suggestions, thanks. Implemented now.

**Review by lmesnik — APPROVED** [review]

**Comment by lmesnik** [comment]
The fix looks good, just one nit.
 rml.wait(100); might be changed to plain sleep(100)

**Comment by sspitsyn** [comment]
> The fix looks good, just one nit. rml.wait(100); might be changed to plain sleep(100)

Right, thanks. Will fix it.

**Comment by sspitsyn** [comment]
Thank you for review, Leonid!

**Review by lmesnik — APPROVED** [review]

**Comment by sspitsyn** [comment]
Ping!! Need one more review, please.

File: test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java [hunk 2]
```diff
@@ -53,9 +53,9 @@ public class ObjectMonitorUsage {
     static Object lockCheck = new Object();
 
     native static int getRes();
-    native static int waitsToEnter();
-    native static int waitsToBeNotified();
     native static int setTestedMonitor(Object monitor);
+    native static void ensureBlockedOnEnter(Thread thread);
+    native static void ensureWaitsToBeNotified(Thread thread);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java:58]: This should be named "Waiting" instead of "Waits".

**Review by plummercj — CHANGES_REQUESTED** [review]

**Review by plummercj — APPROVED** [review]
Looks good.

**Comment by sspitsyn** [comment]
Thank you for review, Leonid and Chris!

**Pull request closed without merging**